sycl/doc/ControllingSYCLLanguageFeatures.md

```diff
@@ -0,0 +1,195 @@
+  corresponding macro. So, if we require from user to specify the macro for an
+  extension, it contradicts with the current proposal of extensions mechanism in
+  SYCL.
+
+- It is inconsistent for users (and they likely don't really care about how
+  particular extension is designed internally) why some of them are enabled
+  via some compiler flag and another ones are enabled via macro. Also,
+  implementation of every extension might be changed in the future as extension
+  and the whole SYCL implementation evolves over time.
+
+- It is easy to make a typo in some extension name and instead of having one
+  clear error that extension name is invalid, user will get bunch of errors that
+  particular types/methods or functions are not available. For example:
+  `-DSYCL_INTEL_SUBGRUOP_ALGORITHMS_ENABLE` - note the typo.
+
+
+[SYCL_INTEL_extension_api]: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/ExtensionMechanism/SYCL_INTEL_extension_api.asciidoc
+[Predefined macros]: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/ExtensionMechanism/SYCL_INTEL_extension_api.asciidoc#predefined-macros
+
+#### -sycl-ext vs. -fsycl-extension-name/-fno-sycl-extension-name
+
+Another one way to enable/disable extensions is to provide separate option
+for each extensions.
+
+There are pros and cons of both approaches:
+
+##### Single flag, i.e. -sycl-ext
+
+Pros:
+- We don't have to hardcode list of known extensions and update the compiler
+  for each new extension we have prepared: `-cl-std` works in this manner
+
+  - This simplifies prototyping of new extensions, especially if they are
+    header-only
+
+- This allows to easily treat header-only extensions in the same way as
+  actual compiler extension and change the implementation without any
+  updates in compiler/toolchain interface (command line options)
+
+- Enabling/disabling extension via this option could automatically define
+  corresponding macro, which can be used by header-only extensions to hide
+  new classes/APIs and other stuff introduced by the extension to avoid it
+  being accidentally used
+
+Cons:
+- Without list of known extensions, we cannot emit proper diagnostic that some
+  unknown extension was enabled just because of the typo in its spelling
+
+  - Potentially, we could introduce one more option, which will allow to check
+    that list of enabled/disabled extensions doesn't contain anything unknown,
+    but this again means change of the compiler for each even header-only
+    extension
+
+- According to [Predefined macros] section from [SYCL_INTEL_extension_api]
+  proposal, extension-specific macro should not only be defined, but also has a
+  value in particular form. How can we automatically put any meaningful value
+  in there without having predefined list of known extensions? Do we need to
+  extend the format so user can specify which version of the extension is
+  needed (`-sycl-ext=+EXTENSION_NAME1=123`)?
```

Comment:
Having a multiple-option flag can be problematic for build systems, or for users that have scripts to generate command line flags. Is there an example of this type of flag currently in clang/llvm ?

Reply:
Documented my intent about several occurrences of `-sycl-ext` in 349ec2df85e9908d0980b818a7c26dafe04d3cf0

Reply:
Trying to use this combination of -sycl-ext accurately on a build system like CMake could be difficult. I don't know enough about it to answer it, maybe I am just being overly cautious. 

```quote
Moreover, the intent is to allow to pass -sycl-ext flag several times: the final list of extensions will be calculated as sum of values for each occurrence of the flag.
```

This can be confusing. Usually on the compiler command line, the latest command is the winner. However, now you have to go and find the entire command line and count all occurences to figure out which extension is enabled and which one is disabled? When you use it directly on the command line may be fine, but when you use this from some generator (again, like CMake, but may as well be some python script), you may end up having some unexpected outcomes.